From 886c804469c448a4c2e422f937f8d147c3e70225 Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@turris.com>
Date: Thu, 24 Aug 2023 08:50:25 +0200
Subject: [PATCH] samba4: Backport from Openwrt 23.05

The version we have currently is outdated and contains security issues.
The easiest way to fix that is to update to the latest community
maintained version.
---
 net/samba4/Config.in                          | 12 ++++-
 net/samba4/Makefile                           | 40 ++++++++++-------
 net/samba4/patches/001-samba-4.4.0-pam.patch  | 29 ------------
 net/samba4/patches/002-dnsserver-4.7.0.patch  | 19 --------
 net/samba4/patches/006-netdb-defines.patch    | 17 -------
 ...bldb-fix-musl-libc-unkown-type-error.patch |  5 +--
 ...mba-4-11-fix-host-tools-checks.patch.patch | 31 ++-----------
 ...-fix-musl_missing__nss_buflen_passwd.patch |  9 ++--
 ...-14-disable-python-module-host-check.patch |  6 +--
 .../020-source3-msgsock-nvram-fix.patch       |  4 +-
 .../021-source4-msgsock-nvram-fix.patch       |  2 +-
 .../102-samba-4.11-unbundle-libbsd.patch      |  8 ++--
 .../103-samba-4.12-unbundle-libunwind.patch   | 44 +++++++++++--------
 net/samba4/patches/105-perl-json-pp.patch     | 20 +++++++++
 net/samba4/waf-cross-answers/aarch64.txt      |  2 +
 net/samba4/waf-cross-answers/arc.txt          |  2 +
 net/samba4/waf-cross-answers/arm.txt          |  2 +
 net/samba4/waf-cross-answers/armeb.txt        |  2 +
 net/samba4/waf-cross-answers/i386.txt         |  2 +
 net/samba4/waf-cross-answers/mips.txt         |  2 +
 net/samba4/waf-cross-answers/mips64.txt       |  2 +
 net/samba4/waf-cross-answers/mips64el.txt     |  2 +
 net/samba4/waf-cross-answers/mipsel.txt       |  2 +
 net/samba4/waf-cross-answers/powerpc.txt      |  2 +
 net/samba4/waf-cross-answers/powerpc64.txt    |  2 +
 net/samba4/waf-cross-answers/riscv64.txt      | 40 +++++++++++++++++
 net/samba4/waf-cross-answers/x86_64.txt       |  2 +
 27 files changed, 166 insertions(+), 144 deletions(-)
 delete mode 100644 net/samba4/patches/001-samba-4.4.0-pam.patch
 delete mode 100644 net/samba4/patches/002-dnsserver-4.7.0.patch
 delete mode 100644 net/samba4/patches/006-netdb-defines.patch
 create mode 100644 net/samba4/patches/105-perl-json-pp.patch
 create mode 100644 net/samba4/waf-cross-answers/riscv64.txt

diff --git a/net/samba4/Config.in b/net/samba4/Config.in
index d287effa5..21cbb1dc2 100644
--- a/net/samba4/Config.in
+++ b/net/samba4/Config.in
@@ -31,12 +31,22 @@ config SAMBA4_SERVER_AVAHI
 		Announce Samba resources via DNS/DNS-SD using the Avahi daemon, for Linux/Mac clients.
 	default y
 
+config SAMBA4_SERVER_QUOTAS
+	bool "Quotas support"
+	depends on PACKAGE_samba4-server
+	select SAMBA4_SERVER_VFS
+	help
+		Enable VFS Quotas
+		installs:
+			modules: vfs_default_quota
+	default n
+
 config SAMBA4_SERVER_VFS
 	bool "Common VFS modules"
 	depends on PACKAGE_samba4-server
 	help
 		installs:
-			modules: (vfs_btrfs) vfs_fruit vfs_shadow_copy2 vfs_recycle vfs_fake_perms vfs_readonly vfs_cap vfs_offline vfs_crossrename vfs_catia vfs_streams_xattr vfs_default_quota
+			modules: (vfs_btrfs) vfs_fruit vfs_shadow_copy2 vfs_recycle vfs_fake_perms vfs_readonly vfs_cap vfs_offline vfs_crossrename vfs_catia vfs_streams_xattr
 
 		Commonly used VFS modules, vfs_btrfs requires kmod-fs-btrfs to be selected separately
 	default y
diff --git a/net/samba4/Makefile b/net/samba4/Makefile
index d84caa0fc..4144ee152 100644
--- a/net/samba4/Makefile
+++ b/net/samba4/Makefile
@@ -2,8 +2,8 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=samba
-PKG_VERSION:=4.14.12
-PKG_RELEASE:=$(AUTORELEASE)
+PKG_VERSION:=4.18.0
+PKG_RELEASE:=1
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:= \
@@ -13,20 +13,22 @@ PKG_SOURCE_URL:= \
 		http://www.nic.funet.fi/index/samba/pub/samba/stable/ \
 		http://samba.mirror.bit.nl/samba/ftp/stable/ \
 		https://download.samba.org/pub/samba/stable/
-PKG_HASH:=155d9c2dfb06a18104422987590858bfe5e9783ebebe63882e7e7f07eaaa512d
+PKG_HASH:=70348656ef807be9c8be4465ca157cef4d99818e234253d2c684cc18b8408149
+
+PKG_BUILD_FLAGS:=gc-sections
 
-PKG_MAINTAINER:=Andy Walsh <andy.walsh44+github@gmail.com>
 PKG_LICENSE:=GPL-3.0-only
 PKG_LICENSE_FILES:=COPYING
 PKG_CPE_ID:=cpe:/a:samba:samba
 
 # samba4=(asn1_compile,compile_et) rpcsvc-proto=(rpcgen)
 HOST_BUILD_DEPENDS:=python3/host rpcsvc-proto/host perl/host perl-parse-yapp/host
-PKG_BUILD_DEPENDS:=samba4/host libtasn1/host
+PKG_BUILD_DEPENDS:=samba4/host libtasn1/host perl/host
 
 PKG_CONFIG_DEPENDS:= \
 	CONFIG_SAMBA4_SERVER_NETBIOS \
 	CONFIG_SAMBA4_SERVER_AVAHI \
+	CONFIG_SAMBA4_SERVER_QUOTAS \
 	CONFIG_SAMBA4_SERVER_VFS \
 	CONFIG_SAMBA4_SERVER_VFSX \
 	CONFIG_SAMBA4_SERVER_AD_DC \
@@ -121,13 +123,13 @@ define Package/samba4-utils
 endef
 
 define Package/samba4-utils/description
-  installs: smbstatus smbtree mvxattr smbtar smbcquotas
+  installs: smbstatus smbtree mvxattr smbtar (smbcquotas)
 
   Utilities collection
 endef
 
-TARGET_CFLAGS += $(FPIC) -ffunction-sections -fdata-sections
-TARGET_LDFLAGS += -Wl,--gc-sections,--as-needed
+TARGET_CFLAGS += $(FPIC)
+TARGET_LDFLAGS += -Wl,--as-needed
 # dont mess with sambas private rpath!
 RSTRIP:=:
 
@@ -178,7 +180,6 @@ CONFIGURE_ARGS += \
 		--without-automount \
 		--without-iconv \
 		--without-lttng \
-		--without-ntvfs-fileserver \
 		--without-pam \
 		--without-systemd \
 		--without-utmp \
@@ -203,7 +204,6 @@ HOST_CONFIGURE_ARGS += \
 		--without-automount \
 		--without-iconv \
 		--without-lttng \
-		--without-ntvfs-fileserver \
 		--without-pam \
 		--without-systemd \
 		--without-utmp \
@@ -215,7 +215,7 @@ HOST_CONFIGURE_ARGS += \
 
 HOST_CONFIGURE_ARGS += --disable-avahi --without-quotas --without-acl-support --without-winbind \
 	--without-ad-dc --without-json --without-libarchive --disable-python --nopyc --nopyo \
-	--without-dnsupdate --without-ads --without-ldap --without-ldb-lmdb
+	--without-ads --without-ldap --without-ldb-lmdb
 
 # Optional AES-NI support - https://lists.samba.org/archive/samba-technical/2017-September/122738.html
 # Support for Nettle wasn't comitted
@@ -232,7 +232,7 @@ CONFIGURE_ARGS += \
 		--with-privatedir=/etc/samba
 
 # features
-ifeq ($(CONFIG_SAMBA4_SERVER_VFS),y)
+ifeq ($(CONFIG_SAMBA4_SERVER_QUOTAS),y)
 	CONFIGURE_ARGS += --with-quotas
 else
 	CONFIGURE_ARGS += --without-quotas
@@ -247,7 +247,7 @@ ifeq ($(CONFIG_SAMBA4_SERVER_AD_DC),y)
 	CONFIGURE_ARGS += --without-winbind --without-ldb-lmdb --with-acl-support
 else
 	CONFIGURE_ARGS += --without-winbind --without-ads --without-ldap --without-ldb-lmdb --without-ad-dc \
-		--without-json --without-libarchive --disable-python --nopyc --nopyo --without-dnsupdate --without-acl-support
+		--without-json --without-libarchive --disable-python --nopyc --nopyo --without-acl-support
 endif
 
 SAMBA4_PDB_MODULES :=pdb_smbpasswd,pdb_tdbsam,
@@ -259,7 +259,10 @@ ifdef CONFIG_KERNEL_IO_URING
 	SAMBA4_VFS_MODULES_SHARED :=$(SAMBA4_VFS_MODULES_SHARED)vfs_io_uring,
 endif
 ifeq ($(CONFIG_SAMBA4_SERVER_VFS),y)
-	SAMBA4_VFS_MODULES_SHARED :=$(SAMBA4_VFS_MODULES_SHARED)vfs_fruit,vfs_shadow_copy2,vfs_recycle,vfs_fake_perms,vfs_readonly,vfs_cap,vfs_offline,vfs_crossrename,vfs_catia,vfs_streams_xattr,vfs_xattr_tdb,vfs_default_quota,vfs_widelinks,
+	SAMBA4_VFS_MODULES_SHARED :=$(SAMBA4_VFS_MODULES_SHARED)vfs_fruit,vfs_shadow_copy2,vfs_recycle,vfs_fake_perms,vfs_readonly,vfs_cap,vfs_offline,vfs_crossrename,vfs_catia,vfs_streams_xattr,vfs_xattr_tdb,vfs_widelinks,
+ifeq ($(CONFIG_SAMBA4_SERVER_QUOTAS),y)
+	SAMBA4_VFS_MODULES_SHARED :=$(SAMBA4_VFS_MODULES_SHARED)vfs_default_quota,
+endif
 ifdef CONFIG_PACKAGE_kmod-fs-btrfs
 	SAMBA4_VFS_MODULES_SHARED :=$(SAMBA4_VFS_MODULES_SHARED)vfs_btrfs,
 endif
@@ -310,6 +313,7 @@ CONFIGURE_ARGS += --private-libraries=$(SYSTEM_PRIVATE_BUNDLED_LIBS)
 
 export COMPILE_ET=$(STAGING_DIR_HOSTPKG)/bin/compile_et_samba
 export ASN1_COMPILE=$(STAGING_DIR_HOSTPKG)/bin/asn1_compile_samba
+export PYTHONHASHSEED=1
 
 # make sure we use the hostpkg build toolset and we need to find host 'yapp'
 HOST_CONFIGURE_VARS+= \
@@ -328,6 +332,9 @@ define Host/Prepare
 	$(call Host/Prepare/Default)
 	$(SED) 's,mandatory=True,mandatory=False,g' $(HOST_BUILD_DIR)/wscript_configure_system_gnutls
 	$(SED) 's,gnutls_version =.*,gnutls_version = gnutls_min_required_version,g' $(HOST_BUILD_DIR)/wscript_configure_system_gnutls
+	$(SED) 's,gnutls_version_str.*,gnutls_version_str = "3.7.7",g' $(HOST_BUILD_DIR)/wscript_configure_system_gnutls
+	$(SED) 's,(gnutls_version > .*,(parse_version(gnutls_version) > parse_version("3.6.10")):,g' $(HOST_BUILD_DIR)/wscript_configure_system_gnutls
+	$(SED) 's,(gnutls_version < .*,(parse_version(gnutls_version) < parse_version("3.5.2")):,g' $(HOST_BUILD_DIR)/wscript_configure_system_gnutls
 endef
 define Host/Compile
 	(cd $(HOST_BUILD_DIR); \
@@ -351,6 +358,9 @@ ifeq ($(CONFIG_SAMBA4_SERVER_AD_DC),)
 	# unbundle iso8601
 	$(SED) '/"iso8601":/d' $(PKG_BUILD_DIR)/third_party/wscript
 endif
+ifeq ($(CONFIG_HOST_OS_MACOS),y)
+	$(SED) "/^VERSION=/i sys.platform='linux'" $(PKG_BUILD_DIR)/buildtools/bin/waf
+endif
 endef
 
 define Build/Configure
@@ -401,7 +411,7 @@ endef
 define Package/samba4-utils/install
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/{smbstatus,smbtree,mvxattr,smbtar} $(1)/usr/bin/
-ifeq ($(CONFIG_SAMBA4_SERVER_VFS),y)
+ifeq ($(CONFIG_SAMBA4_SERVER_QUOTAS),y)
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/smbcquotas $(1)/usr/bin/
 endif
 endef
diff --git a/net/samba4/patches/001-samba-4.4.0-pam.patch b/net/samba4/patches/001-samba-4.4.0-pam.patch
deleted file mode 100644
index 72de06619..000000000
--- a/net/samba4/patches/001-samba-4.4.0-pam.patch
+++ /dev/null
@@ -1,29 +0,0 @@
---- a/source3/wscript
-+++ b/source3/wscript
-@@ -885,7 +885,7 @@ msg.msg_accrightslen = sizeof(fd);
-         if conf.env.with_iconv:
-             conf.DEFINE('HAVE_ICONV', 1)
- 
--    if Options.options.with_pam:
-+    if Options.options.with_pam != False:
-         use_pam=True
-         conf.CHECK_HEADERS('security/pam_appl.h pam/pam_appl.h')
-         if not conf.CONFIG_SET('HAVE_SECURITY_PAM_APPL_H') and not conf.CONFIG_SET('HAVE_PAM_PAM_APPL_H'):
-@@ -962,6 +962,17 @@ int i; i = PAM_RADIO_TYPE;
-                        "or headers not found. Use --without-pam to disable "
-                        "PAM support.");
- 
-+    else:
-+        Logs.warn("PAM disabled")
-+        use_pam=False
-+        conf.undefine('WITH_PAM')
-+        conf.undefine('WITH_PAM_MODULES')
-+        conf.undefine('HAVE_SECURITY_PAM_APPL_H')
-+        conf.undefine('PAM_RHOST')
-+        conf.undefine('PAM_TTY')
-+        conf.undefine('HAVE_PAM_PAM_APPL_H')
-+
-+
-     seteuid = False
- 
- #
diff --git a/net/samba4/patches/002-dnsserver-4.7.0.patch b/net/samba4/patches/002-dnsserver-4.7.0.patch
deleted file mode 100644
index 0c7592f8e..000000000
--- a/net/samba4/patches/002-dnsserver-4.7.0.patch
+++ /dev/null
@@ -1,19 +0,0 @@
-samba: build dnsserver_common code
-
-Just 'install' does not seem to do it.
-
-Upstream-Status: Pending
-
-Signed-off-by: Joe Slater <joe.slater@windriver.com>
-
---- a/source4/dns_server/wscript_build
-+++ b/source4/dns_server/wscript_build
-@@ -4,7 +4,7 @@ bld.SAMBA_LIBRARY('dnsserver_common',
-         source='dnsserver_common.c',
-         deps='samba-util samba-errors ldbsamba clidns',
-         private_library=True,
--        install=bld.AD_DC_BUILD_IS_ENABLED()
-+        enabled=bld.AD_DC_BUILD_IS_ENABLED()
-         )
- 
- bld.SAMBA_MODULE('service_dns',
diff --git a/net/samba4/patches/006-netdb-defines.patch b/net/samba4/patches/006-netdb-defines.patch
deleted file mode 100644
index a11f8d15c..000000000
--- a/net/samba4/patches/006-netdb-defines.patch
+++ /dev/null
@@ -1,17 +0,0 @@
---- a/nsswitch/wins.c
-+++ b/nsswitch/wins.c
-@@ -40,6 +40,14 @@ static pthread_mutex_t wins_nss_mutex =
- #define INADDRSZ 4
- #endif
- 
-+#ifndef NETDB_INTERNAL
-+#define NETDB_INTERNAL -1
-+#endif
-+
-+#ifndef NETDB_SUCCESS
-+#define NETDB_SUCCESS 0
-+#endif
-+
- NSS_STATUS _nss_wins_gethostbyname_r(const char *hostname,
- 				     struct hostent *he,
- 				     char *buffer,
diff --git a/net/samba4/patches/007-libldb-fix-musl-libc-unkown-type-error.patch b/net/samba4/patches/007-libldb-fix-musl-libc-unkown-type-error.patch
index 29928976b..da564e050 100644
--- a/net/samba4/patches/007-libldb-fix-musl-libc-unkown-type-error.patch
+++ b/net/samba4/patches/007-libldb-fix-musl-libc-unkown-type-error.patch
@@ -15,12 +15,11 @@ Signed-off-by: Changqing Li <changqing.li@windriver.com>
 
 --- a/lib/tevent/tevent.h
 +++ b/lib/tevent/tevent.h
-@@ -32,6 +32,8 @@
- #include <talloc.h>
+@@ -33,6 +33,7 @@
  #include <sys/time.h>
+ #include <sys/types.h>
  #include <stdbool.h>
 +#include <sys/stat.h>
-+#include <sys/types.h>
  
  /* for old gcc releases that don't have the feature test macro __has_attribute */
  #ifndef __has_attribute
diff --git a/net/samba4/patches/009-samba-4-11-fix-host-tools-checks.patch.patch b/net/samba4/patches/009-samba-4-11-fix-host-tools-checks.patch.patch
index 284c09d85..4f06db369 100644
--- a/net/samba4/patches/009-samba-4-11-fix-host-tools-checks.patch.patch
+++ b/net/samba4/patches/009-samba-4-11-fix-host-tools-checks.patch.patch
@@ -1,9 +1,9 @@
 --- a/wscript_configure_embedded_heimdal
 +++ b/wscript_configure_embedded_heimdal
-@@ -2,3 +2,14 @@ if not conf.env['FLEX']:
-     conf.fatal("Embedded Heimdal build requires flex but it was not found.  Install flex or use --with-system-mitkrb5 or --with-system-heimdalkrb5")
+@@ -6,3 +6,14 @@ if not conf.env['BISON']:
  
- conf.RECURSE('source4/heimdal_build')
+ conf.define('USING_EMBEDDED_HEIMDAL', 1)
+ conf.RECURSE('third_party/heimdal_build')
 +
 +def check_system_heimdal_binary(name):
 +    if conf.LIB_MAY_BE_BUNDLED(name):
@@ -15,28 +15,3 @@
 +
 +check_system_heimdal_binary("compile_et")
 +check_system_heimdal_binary("asn1_compile")
---- a/wscript_configure_system_heimdal
-+++ b/wscript_configure_system_heimdal
-@@ -37,14 +37,6 @@ def check_system_heimdal_lib(name, funct
-     conf.define('USING_SYSTEM_%s' % name.upper(), 1)
-     return True
- 
--def check_system_heimdal_binary(name):
--    if conf.LIB_MAY_BE_BUNDLED(name):
--        return False
--    if not conf.find_program(name, var=name.upper()):
--        return False
--    conf.define('USING_SYSTEM_%s' % name.upper(), 1)
--    return True
--
- check_system_heimdal_lib("com_err", "com_right_r com_err", "com_err.h")
- 
- if check_system_heimdal_lib("roken", "rk_socket_set_reuseaddr", "roken.h"):
-@@ -96,7 +88,4 @@ finally:
- #if conf.CHECK_BUNDLED_SYSTEM('tommath', checkfunctions='mp_init', headers='tommath.h'):
- #    conf.define('USING_SYSTEM_TOMMATH', 1)
- 
--check_system_heimdal_binary("compile_et")
--check_system_heimdal_binary("asn1_compile")
--
- conf.define('USING_SYSTEM_KRB5', 1)
diff --git a/net/samba4/patches/010-samba-4-12-fix-musl_missing__nss_buflen_passwd.patch b/net/samba4/patches/010-samba-4-12-fix-musl_missing__nss_buflen_passwd.patch
index bbc66a057..b0d08f5c8 100644
--- a/net/samba4/patches/010-samba-4-12-fix-musl_missing__nss_buflen_passwd.patch
+++ b/net/samba4/patches/010-samba-4-12-fix-musl_missing__nss_buflen_passwd.patch
@@ -1,12 +1,13 @@
 --- a/lib/util/util_paths.c
 +++ b/lib/util/util_paths.c
-@@ -26,6 +26,9 @@
- #include "dynconfig/dynconfig.h"
- #include "lib/util/util_paths.h"
+@@ -28,6 +28,10 @@
  #include "system/passwd.h"
+ #include "system/filesys.h"
+ 
 +#if !defined(__GLIBC__) 
 +	#define NSS_BUFLEN_PASSWD 1024
 +#endif
- 
++
  /**
   * @brief Returns an absolute path to a file in the Samba modules directory.
+  *
diff --git a/net/samba4/patches/011-samba-4-14-disable-python-module-host-check.patch b/net/samba4/patches/011-samba-4-14-disable-python-module-host-check.patch
index ee84f779d..d28f08b9d 100644
--- a/net/samba4/patches/011-samba-4-14-disable-python-module-host-check.patch
+++ b/net/samba4/patches/011-samba-4-14-disable-python-module-host-check.patch
@@ -1,8 +1,8 @@
 --- a/python/wscript
 +++ b/python/wscript
-@@ -73,9 +73,9 @@ def configure(conf):
-         for module, package in selftest_pkgs.items():
-             find_third_party_module(conf, module, package)
+@@ -90,9 +90,9 @@ def configure(conf):
+                                       "'python3-iso8601'. Please install "
+                                       "one of the packages.")
  
 -    if not Options.options.without_ad_dc:
 -        for module, package in ad_dc_pkgs.items():
diff --git a/net/samba4/patches/020-source3-msgsock-nvram-fix.patch b/net/samba4/patches/020-source3-msgsock-nvram-fix.patch
index 351f87bf7..df5f56348 100644
--- a/net/samba4/patches/020-source3-msgsock-nvram-fix.patch
+++ b/net/samba4/patches/020-source3-msgsock-nvram-fix.patch
@@ -1,6 +1,6 @@
 --- a/source3/lib/messages.c
 +++ b/source3/lib/messages.c
-@@ -507,7 +507,7 @@ static NTSTATUS messaging_init_internal(
+@@ -500,7 +500,7 @@ static NTSTATUS messaging_init_internal(
  		return NT_STATUS_ACCESS_DENIED;
  	}
  
@@ -9,7 +9,7 @@
  	if (priv_path == NULL) {
  		return NT_STATUS_NO_MEMORY;
  	}
-@@ -670,7 +670,7 @@ NTSTATUS messaging_reinit(struct messagi
+@@ -663,7 +663,7 @@ NTSTATUS messaging_reinit(struct messagi
  		msg_ctx->per_process_talloc_ctx,
  		msg_ctx->event_ctx,
  		&msg_ctx->id.unique_id,
diff --git a/net/samba4/patches/021-source4-msgsock-nvram-fix.patch b/net/samba4/patches/021-source4-msgsock-nvram-fix.patch
index d05e7bdcc..2bd1e79c6 100644
--- a/net/samba4/patches/021-source4-msgsock-nvram-fix.patch
+++ b/net/samba4/patches/021-source4-msgsock-nvram-fix.patch
@@ -1,6 +1,6 @@
 --- a/source4/lib/messaging/messaging.c
 +++ b/source4/lib/messaging/messaging.c
-@@ -500,7 +500,7 @@ struct imessaging_context *imessaging_in
+@@ -526,7 +526,7 @@ static struct imessaging_context *imessa
  		goto fail;
  	}
  
diff --git a/net/samba4/patches/102-samba-4.11-unbundle-libbsd.patch b/net/samba4/patches/102-samba-4.11-unbundle-libbsd.patch
index 00c69bcc5..766fefc52 100644
--- a/net/samba4/patches/102-samba-4.11-unbundle-libbsd.patch
+++ b/net/samba4/patches/102-samba-4.11-unbundle-libbsd.patch
@@ -1,6 +1,6 @@
 --- a/lib/replace/wscript
 +++ b/lib/replace/wscript
-@@ -416,22 +416,13 @@ def configure(conf):
+@@ -434,22 +434,13 @@ def configure(conf):
  
      conf.CHECK_FUNCS('prctl dirname basename')
  
@@ -29,7 +29,7 @@
  
      conf.CHECK_CODE('''
                  struct ucred cred;
-@@ -817,9 +808,6 @@ syscall(SYS_copy_file_range,0,NULL,0,NUL
+@@ -832,9 +823,6 @@ syscall(SYS_copy_file_range,0,NULL,0,NUL
  
      # look for a method of finding the list of network interfaces
      for method in ['HAVE_IFACE_GETIFADDRS', 'HAVE_IFACE_AIX', 'HAVE_IFACE_IFCONF', 'HAVE_IFACE_IFREQ']:
@@ -39,7 +39,7 @@
          if conf.CHECK_CODE('''
                             #define %s 1
                             #define NO_CONFIG_H 1
-@@ -832,7 +820,7 @@ syscall(SYS_copy_file_range,0,NULL,0,NUL
+@@ -847,7 +835,7 @@ syscall(SYS_copy_file_range,0,NULL,0,NUL
                             #include "tests/getifaddrs.c"
                             ''' % method,
                             method,
@@ -48,7 +48,7 @@
                             addmain=False,
                             execute=True):
              break
-@@ -880,7 +868,6 @@ def build(bld):
+@@ -895,7 +883,6 @@ def build(bld):
                  break
  
      extra_libs = ''
diff --git a/net/samba4/patches/103-samba-4.12-unbundle-libunwind.patch b/net/samba4/patches/103-samba-4.12-unbundle-libunwind.patch
index 14a010c42..e680e91ba 100644
--- a/net/samba4/patches/103-samba-4.12-unbundle-libunwind.patch
+++ b/net/samba4/patches/103-samba-4.12-unbundle-libunwind.patch
@@ -1,23 +1,31 @@
 --- a/lib/util/wscript_configure
 +++ b/lib/util/wscript_configure
-@@ -11,13 +11,13 @@ conf.CHECK_FUNCS_IN('backtrace backtrace
+@@ -12,17 +12,17 @@ conf.CHECK_FUNCS_IN('backtrace backtrace
  conf.CHECK_HEADERS('execinfo.h')
  
  conf.SET_TARGET_TYPE('LIBUNWIND', 'EMPTY')
--if conf.check_cfg(package='libunwind-generic',
--                  args='--cflags --libs',
--                  msg='Checking for libunwind',
--                  uselib_store='LIBUNWIND',
--                  mandatory=False):
--    if conf.CHECK_HEADERS('libunwind.h'):
--        conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
-+# if conf.check_cfg(package='libunwind-generic',
-+                  # args='--cflags --libs',
-+                  # msg='Checking for libunwind',
-+                  # uselib_store='LIBUNWIND',
-+                  # mandatory=False):
-+    # if conf.CHECK_HEADERS('libunwind.h'):
-+        # conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
- 
- conf.CHECK_STRUCTURE_MEMBER('struct statvfs', 'f_frsize', define='HAVE_FRSIZE', headers='sys/statvfs.h')
- 
+-if Options.options.with_libunwind:
+-    if conf.check_cfg(package='libunwind-generic',
+-                      args='--cflags --libs',
+-                      msg='Checking for libunwind',
+-                      uselib_store='LIBUNWIND',
+-                      mandatory=False):
+-        if conf.CHECK_HEADERS('libunwind.h'):
+-            conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
+-    else:
+-        raise Errors.WafError('--with-libunwind specified but libunwind not found')
+-elif Options.options.with_libunwind == None:
++#if Options.options.with_libunwind:
++    #if conf.check_cfg(package='libunwind-generic',
++                      #args='--cflags --libs',
++                      #msg='Checking for libunwind',
++                      #uselib_store='LIBUNWIND',
++                      #mandatory=False):
++        #if conf.CHECK_HEADERS('libunwind.h'):
++            #conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
++    #else:
++        #raise Errors.WafError('--with-libunwind specified but libunwind not found')
++if Options.options.with_libunwind == None:
+     if not conf.CONFIG_SET('HAVE_BACKTRACE_SYMBOLS') \
+        and not Options.options.disable_fault_handling:
+         raise Errors.WafError(
diff --git a/net/samba4/patches/105-perl-json-pp.patch b/net/samba4/patches/105-perl-json-pp.patch
new file mode 100644
index 000000000..feb8aaaf4
--- /dev/null
+++ b/net/samba4/patches/105-perl-json-pp.patch
@@ -0,0 +1,20 @@
+--- a/third_party/heimdal/cf/make-proto.pl
++++ b/third_party/heimdal/cf/make-proto.pl
+@@ -4,7 +4,7 @@
+ use Getopt::Std;
+ use File::Compare;
+ 
+-use JSON;
++use JSON::PP;
+ 
+ my $comment = 0;
+ my $doxygen = 0;
+@@ -70,7 +70,7 @@ if($opt_x) {
+     my $EXP;
+     local $/;
+     open(EXP, '<', $opt_x) || die "open ${opt_x}";
+-    my $obj = JSON->new->utf8->decode(<EXP>);
++    my $obj = JSON::PP->new->utf8->decode(<EXP>);
+     close $EXP;
+ 
+     foreach my $x (keys %$obj) {
diff --git a/net/samba4/waf-cross-answers/aarch64.txt b/net/samba4/waf-cross-answers/aarch64.txt
index 5d9af17e1..0714416f8 100644
--- a/net/samba4/waf-cross-answers/aarch64.txt
+++ b/net/samba4/waf-cross-answers/aarch64.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: OK
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/arc.txt b/net/samba4/waf-cross-answers/arc.txt
index 0d57fbc3a..e0635f7a8 100644
--- a/net/samba4/waf-cross-answers/arc.txt
+++ b/net/samba4/waf-cross-answers/arc.txt
@@ -39,3 +39,5 @@ getcwd takes a NULL argument: OK
 Checking whether setreuid is available: NO
 Checking whether setresuid is available: NO
 Checking whether seteuid is available: NO
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/arm.txt b/net/samba4/waf-cross-answers/arm.txt
index df5f77bbc..9d07b6f27 100644
--- a/net/samba4/waf-cross-answers/arm.txt
+++ b/net/samba4/waf-cross-answers/arm.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/armeb.txt b/net/samba4/waf-cross-answers/armeb.txt
index df5f77bbc..9d07b6f27 100644
--- a/net/samba4/waf-cross-answers/armeb.txt
+++ b/net/samba4/waf-cross-answers/armeb.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/i386.txt b/net/samba4/waf-cross-answers/i386.txt
index df5f77bbc..9d07b6f27 100644
--- a/net/samba4/waf-cross-answers/i386.txt
+++ b/net/samba4/waf-cross-answers/i386.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/mips.txt b/net/samba4/waf-cross-answers/mips.txt
index 04b93e589..e88772111 100644
--- a/net/samba4/waf-cross-answers/mips.txt
+++ b/net/samba4/waf-cross-answers/mips.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/mips64.txt b/net/samba4/waf-cross-answers/mips64.txt
index 86be3763b..73f77af37 100644
--- a/net/samba4/waf-cross-answers/mips64.txt
+++ b/net/samba4/waf-cross-answers/mips64.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: OK
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/mips64el.txt b/net/samba4/waf-cross-answers/mips64el.txt
index 86be3763b..73f77af37 100644
--- a/net/samba4/waf-cross-answers/mips64el.txt
+++ b/net/samba4/waf-cross-answers/mips64el.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: OK
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/mipsel.txt b/net/samba4/waf-cross-answers/mipsel.txt
index 04b93e589..e88772111 100644
--- a/net/samba4/waf-cross-answers/mipsel.txt
+++ b/net/samba4/waf-cross-answers/mipsel.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/powerpc.txt b/net/samba4/waf-cross-answers/powerpc.txt
index 3f1b18fd4..85a105acb 100644
--- a/net/samba4/waf-cross-answers/powerpc.txt
+++ b/net/samba4/waf-cross-answers/powerpc.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: NO
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/powerpc64.txt b/net/samba4/waf-cross-answers/powerpc64.txt
index 5d9af17e1..0714416f8 100644
--- a/net/samba4/waf-cross-answers/powerpc64.txt
+++ b/net/samba4/waf-cross-answers/powerpc64.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: OK
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/riscv64.txt b/net/samba4/waf-cross-answers/riscv64.txt
new file mode 100644
index 000000000..0714416f8
--- /dev/null
+++ b/net/samba4/waf-cross-answers/riscv64.txt
@@ -0,0 +1,40 @@
+Checking uname sysname type: "Linux"
+Checking simple C program: "hello world"
+rpath library support: OK
+-Wl,--version-script support: OK
+Checking getconf LFS_CFLAGS: NO
+Checking for large file support without additional flags: OK
+Checking correct behavior of strtoll: OK
+Checking for working strptime: NO
+Checking for C99 vsnprintf: "1"
+Checking for HAVE_SHARED_MMAP: OK
+Checking for HAVE_MREMAP: OK
+Checking for HAVE_INCOHERENT_MMAP: NO
+Checking for HAVE_SECURE_MKSTEMP: OK
+Checking for HAVE_IFACE_GETIFADDRS: OK
+Checking value of NSIG: "65"
+Checking value of _NSIG: "65"
+Checking value of SIGRTMAX: "64"
+Checking value of SIGRTMIN: "35"
+Checking for a 64-bit host to support lmdb: OK
+Checking value of GNUTLS_CIPHER_AES_128_CFB8: "29"
+Checking value of GNUTLS_MAC_AES_CMAC_128: "203"
+Checking errno of iconv for illegal multibyte sequence: OK
+Checking for kernel change notify support: OK
+Checking for Linux kernel oplocks: OK
+Checking for kernel share modes: OK
+Checking whether POSIX capabilities are available: OK
+Checking if can we convert from CP850 to UCS-2LE: OK
+Checking if can we convert from UTF-8 to UCS-2LE: OK
+vfs_fileid checking for statfs() and struct statfs.f_fsid: OK
+Checking whether we can use Linux thread-specific credentials: OK
+Checking whether fcntl locking is available: OK
+Checking whether fcntl lock supports open file description locks: NO
+Checking whether fcntl supports flags to send direct I/O availability signals: OK
+Checking whether fcntl supports setting/geting hints: (-11, "")
+Checking for the maximum value of the 'time_t' type: OK
+Checking whether the realpath function allows a NULL argument: OK
+Checking for ftruncate extend: OK
+getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
diff --git a/net/samba4/waf-cross-answers/x86_64.txt b/net/samba4/waf-cross-answers/x86_64.txt
index 5d9af17e1..0714416f8 100644
--- a/net/samba4/waf-cross-answers/x86_64.txt
+++ b/net/samba4/waf-cross-answers/x86_64.txt
@@ -36,3 +36,5 @@ Checking for the maximum value of the 'time_t' type: OK
 Checking whether the realpath function allows a NULL argument: OK
 Checking for ftruncate extend: OK
 getcwd takes a NULL argument: OK
+Checking for readlink breakage: OK
+Checking for gnutls fips mode support: NO
\ No newline at end of file
-- 
2.41.0

